rust-runtime/aws-smithy-client/src/builder.rs

```diff
@@ -260,38 +265,39 @@ impl<C, M, R> Builder<C, M, R> {
         Builder {
             connector: self.connector,
             middleware: map(self.middleware),
-            standard_retry_config: self.standard_retry_config,
-            retry_policy: self.retry_policy,
+            retry_config: self.retry_config,
             timeout_config: self.timeout_config,
             sleep_impl: self.sleep_impl,
         }
     }
 
     /// Build a Smithy service [`Client`].
     pub fn build(self) -> Client<C, M, R> {
+        let timeout_config = self.timeout_config.unwrap_or_else(TimeoutConfig::disabled);
         if self.sleep_impl.is_none() {
             const ADDITIONAL_HELP: &str =
                 "Either disable retry by setting max attempts to one, or pass in a `sleep_impl`. \
                 If you're not using Tokio, then an implementation of the `AsyncSleep` trait from \
                 the `aws-smithy-async` crate is required for your async runtime. If you are using \
                 Tokio, then make sure the `rt-tokio` feature is enabled to have its sleep \
                 implementation set automatically.";
-            if self
-                .standard_retry_config
-                .map(|src| src.has_retry())
-                .unwrap_or(false)
-            {
-                panic!("Retries require a `sleep_impl`, but none was passed into the builder. {ADDITIONAL_HELP}");
-            } else if self.timeout_config.has_timeouts() {
+            if let BuilderRetryConfig::Standard { has_retry, .. } = self.retry_config {
+                if has_retry {
```

Comment:
I'm wondering if we should make this concept more generic and rename it to something like `depends_on_sleep_impl` or `requires_sleep_impl` because timeouts also have similar language. Perhaps we could even make trait (although that could be going too far.)

Reply:
It is now more generic and called `MaybeRequiresSleep` :smile: